Sources_v3/StreamChatUI/ChatChannel/ChatMessageListVC.swift

```diff
@@ -0,0 +1,333 @@
+    }
+
+    public var dataSource: DataSource = .empty()
+    public var delegate: Delegate? // swiftlint:disable:this weak_delegate
+
+    public private(set) lazy var router = uiConfig.navigation.messageListRouter.init(rootViewController: self)
+
+    public private(set) lazy var collectionViewLayout: ChatChannelCollectionViewLayout = uiConfig
+        .messageList
+        .collectionLayout
+        .init()
+    public private(set) lazy var collectionView: UICollectionView = {
+        let collection = uiConfig.messageList.collectionView.init(layout: collectionViewLayout)
+        collection.register(
+            СhatIncomingMessageCollectionViewCell<ExtraData>.self,
+            forCellWithReuseIdentifier: СhatIncomingMessageCollectionViewCell<ExtraData>.reuseId
+        )
+        collection.register(
+            СhatOutgoingMessageCollectionViewCell<ExtraData>.self,
+            forCellWithReuseIdentifier: СhatOutgoingMessageCollectionViewCell<ExtraData>.reuseId
+        )
+        collection.showsHorizontalScrollIndicator = false
+        collection.dataSource = self
+        collection.delegate = self
+
+        return collection
+    }()
+
+    // MARK: - Life Cycle
+
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+
+        let longPress = UILongPressGestureRecognizer(target: self, action: #selector(didLongPress(_:)))
+        collectionView.addGestureRecognizer(longPress)
+    }
+
+    override open func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        if collectionView.contentOffset.y < collectionView.bounds.height {
+            let bottom = collectionView.contentSize.height
+            collectionView.scrollRectToVisible(CGRect(x: 0, y: bottom - 1, width: 1, height: 1), animated: false)
+        }
+    }
+
+    override open func setUpLayout() {
+        super.setUpLayout()
+
+        view.addSubview(collectionView)
+        collectionView.pin(to: view.safeAreaLayoutGuide)
+    }
+
+    override public func defaultAppearance() {
+        super.defaultAppearance()
+
+        view.backgroundColor = uiConfig.colorPalette.generalBackground
+        collectionView.backgroundColor = .clear
+    }
+
+    // MARK: - Public API
+
+    public func updateMessages(with changes: [ListChange<_ChatMessage<ExtraData>>], completion: ((Bool) -> Void)? = nil) {
+        collectionView.performBatchUpdates({
+            for change in changes {
+                switch change {
+                case let .insert(_, index):
+                    collectionView.insertItems(at: [index])
+                case let .move(_, fromIndex, toIndex):
+                    collectionView.moveItem(at: fromIndex, to: toIndex)
+                case let .remove(_, index):
+                    collectionView.deleteItems(at: [index])
+                case let .update(_, index):
+                    collectionView.reloadItems(at: [index])
+                }
+            }
+        }, completion: completion)
```

Comment:
nit: Should we introduce some kind of helper for this boilerplate? Also, I think the reloading should be probably done directly by changing the content of the cell if the cell is visible, not by reloading it completely since we don't change the cell type. This will remove the ugly fade animation.

Reply:
I think it's definitely something we could work on later 👍🏾